### ISSUE TITLE
Tesseract v3.05.00 Fails to Build on OS X/macOS 'El Capitan' v10.11.6 With Options Specified in Issue Body

### ISSUE THREAD

**Issue by RandomDSdevel** [Homebrew/homebrew-core#10380]
### Please follow the general troubleshooting steps first:

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [x] Confirmed this is problem with specific formulae and not Homebrew/brew? If it's a general Homebrew/brew problem please file this issue at https://github.com/Homebrew/brew/issues/new

### Bug reports:

The options involved were:  

* `--with-all-languages`
* `--with-opencl`
* `--with-serial-num-pack`
* `--with-training-tools`

[Here](https://github.com/RandomDSdevel/Homebrew-Installation-Logs/tree/master/Tesseract/v3.05.00/First%20Attempt) are my logs.  Something odd I noticed with the verbose and debug output is that `./autogen.sh` couldn't find `libtoolize`, but I'm not entirely sure what to make of that.  Also, the problem that finally crashed `brew` seems to be related to Tesseract's OpenCL support, which was a bit buggy at run-time even in the previous version despite building well enough (maybe this'll result in an upstream bug report…?  Eh, who knows…)  Finally, I'm currently running a second attempt, but with `HOMEBREW_MAKE_JOBS` set equal to `1` to see if this isn't a build artifact order dependency issue, so I'll report back on whether _that_ crashes or not in a little while.

**Comment by RandomDSdevel** [comment]
OK, the attempt with `HOMEBREW_MAKE_JOBS=1` failed as well, so the problem isn't a build artifact order dependency issue.  [Here](https://github.com/RandomDSdevel/Homebrew-Installation-Logs/tree/master/Tesseract/v3.05.00/Second%20Attempt) are the relevant logs, but you can _probably_ just ignore them, as they're likely nearly, if not _exactly_, identical to the first set.

**Comment by JCount** [comment]
I can confirm that `tesseract` seems to be broken if trying to build `with-opencl`. I'll try and look into it.

**Comment by RandomDSdevel** [comment]
Thanks, @JCount, as I don't think I have the relevant expertise (_yet_, anyway…)

**Comment by JCount** [comment]
@RandomDSdevel I can't figure out what changed with the new release that has broken the OpenCL option. At this point it should be reported upstream, as it seems to be more of an issue on their end. Barring that, input from them regarding what could be going wrong in general would be helpful. If you could report the OpenCL issue you found upstream, it would be quite helpful. Thanks. 

Also, I am sorry I can't give you a solution to this issue. 😞

**Comment by RandomDSdevel** [comment]
That's fine, @JCount.  I'll go ahead and report this issue upstream, then.

**Comment by Tristor** [comment]
@RandomDSdevel:  Can you provide the bug number here from upstream that you reported?  I'm investigating and going to see if I can fix this, as this build is breaking for me as well.  No promises, but might as well dig in :)

**Comment by Tristor** [comment]
It's actually compiling just fine, it's failing to link, FYI.  
```
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by Tristor** [comment]
Looks like it's caused by this.  https://github.com/tesseract-ocr/tesseract/issues/739  There was a patch against 4.x that fixed this, but was not backported to 3.x  https://github.com/tesseract-ocr/tesseract/commit/9ec0c4fa9c983

It will have to be fixed upstream.  I was looking at monkey-patching it, but I don't think that's the right answer.

The recipe may need to be updated to ensure that `libtiff` is a dependency when you install `--with-opencl` though, as this might be required for compatibility once the upstream change is made.




Regarding the error in your logs from autogen.sh, it's a red herring.

From autogen.sh
```
echo "Running libtoolize"
libtoolize -f -c || glibtoolize -f -c || bail_out
libtoolize --automake || glibtoolize --automake || bail_out
```
This is first saying `./autogen.sh: line 74: libtoolize: command not found` because rather than testing for the existence, they just invoke the command.  On OS X there's an Apple-specific version of `libtool` for Mach-O files, which requires the `libtool` from GNU to be installed as `glibtool` and `glibtoolize`.

**Comment by RandomDSdevel** [comment]
@Tristor:  Sorry for the delay.  I hadn't gotten around to reporting this upstream yet because I wasn't entirely sure if I understood it well enough to communicate it effectively as a full bug report and was considering asking somebody involved with Tesseract development if they could look into this via some other, more informal method where that wouldn't be nearly as much of a problem&thinsp;a mailing list, say, if there was one.  Now that we know that my problem here is just a manifestation of a pre-existing bug and that the other issue is unrelated, I'll go ahead and comment upstream.

**Comment by RandomDSdevel** [comment]
@Tristor:  As you can see by the fact that GitHub's issue-mentioning notification system kicked in, I've now commented upstream.  

P. S.:  Extending the `tesseract` formula to include an optional dependency on `libtiff` when built `--with-opencl` actually doesn't sound too hard despite how I've only messed around with version bumps and patching before, so I _might_ go ahead and try my hand at submitting a PR to resolve the aspects of this issue which have to do with how Homebrew handles packages myself once the corresponding problem is fixed upstream.  

P. P. S.:  Heck, I should probably go ahead and tip upstream off about the `libtoolize` issue, too, by opening a separate issue for it there now that you've made what's going on in _that_ respect obvious, too…

**Comment by Tristor** [comment]
Thanks for the update.  Since I wasn't able to fully resolve the issue I didn't pursue filing a PR for the other changes, but if it helps I had already patched them locally as part of my troubleshooting.

The change in the recipe would look something like this:
```ruby
if build.with? "opencl"
  depends_on "libtiff"
end
```

The change in the `autogen.sh` script would look something like this.  First, before we started calling utilizes we should find and capture the absolute path to a working copy of `libtoolize` in a variable, then call it as the variable below.

```bash
# Suppress errors by testing before running libtoolize.
if [ -x $(which libtoolize) ]; then
  $LIBTOOLIZE="$(which libtoolize)"
elif [ -x $(which glibtoolize) ]; then
  $LIBTOOLIZE="$(which glibtoolize)"
else
  echo "Unable to find a valid copy of libtoolize in your PATH"
  bail_out
fi
```

Then replace the section that calls it with:
```bash
echo "Running libtoolize"
$LIBTOOLIZE -f -c || bail_out
$LIBTOOLIZE --automake || bail_out
```

Not sure if they care about the `autogen.sh` bug though as it doesn't actually break anything.  It looks like they dirty-hacked it in so it'd work on OSX when it was originally assuming a GNU userland.

**Comment by RandomDSdevel** [comment]
@Tristor:  You're welcome, and, now that you've shown them to me, I'm confident I can make the changes you've described quite easily.

**Comment by Tristor** [comment]
I just edited my previous comment to modify the test for `libtoolize` to use `command` BASH built-in instead of calling `which`, because the behavior of which was unexpected in that context.  I could probably have fixed it to use `which`, but it's technically more correct to use the shell built-in instead of calling an external utility.  I've tested and verified this works :)

**Comment by Tristor** [comment]
So, in the interest of poking this some more, I grabbed the source tarball from the `tesseract` project, manually modified `api/Makefile.am` and `autogen.sh` and built using the same configure flags that Homebrew provides and it still fails.  It looks like despite configure finding the headers for `libtiff` that it doesn't include the `libtiff` path in `CPPFLAGS` or `LIBS` inside the `Makefile` that gets generated.

I've manually added the correct paths for libs and includes to the `Makefile` and am seeing if it works now.  But there may be something more broken upstream... I'm not super familiar with the tesseract project, this is just a dependency for something else I'm using, so I might be missing something.

**Comment by RandomDSdevel** [comment]
Yeah, me neither.  (Oh, and thanks for the comment fix, too, as that should come in handy.  Again, sorry for any delays incurred by the fact that I'm doing a bit of multitasking down here on my end.)

**Comment by Tristor** [comment]
I haven't read through the `Makefile` line by line, but it's totally ignoring the additions I made to `$LIBS` and `$CPPFLAGS` as it never puts the include when it calls `libtool`.  Obviously, since `libtool` doesn't include `libtiff`, it fails during linking as we see.  Doesn't make sense to me as after changing `api/Makefile.am` it now finds the `libtiff` headers during configure... just ignores them in the future completely.

I will probably comment upstream here in a little bit with the information and hope somebody knows more.

**Comment by RandomDSdevel** [comment]
Yeah, sounds complicated.  It might do my sanity some good to just stick with fixing the `tesseract` formula down here in `brew`-land…

**Comment by RandomDSdevel** [comment]
@Tristor:  Hmm, `tesseract.rb` says `libtiff` is already a `:recommended` optional dependency of Tesseract on [it's 34<sup>th</sup> line](https://github.com/homebrew/homebrew-core/blob/master/Formula/tesseract.rb#L34).  Could I possibly change bump it up to being a required one when building `tesseract` `--with-opencl`?  If so, then what might the syntax for that be?

**Comment by RandomDSdevel** [comment]
@Tristor:  Now that I've thought about it some more, it seems to me that maybe something like:  

```
if build.with? "opencl"
  depends_on "libtiff"
else
  depends_on "libtiff" => :recommended
end
```

might do the trick, or perhaps:  

```
depends_on "libtiff" => ["with-opencl", :recommended]
```

might also work, but that I'm less sure about since I know neither if `:recommended` can be used within option sets nor whether this would be equivalent to my other idea (by allowing `:recommended` to be honored separately from `"with-opencl"`, that is.)  Also, I was planning as mentioning the pull request implementing this change as depending on upstream changes in its initial description anyway, so that should minimize confusion as to when to merge it&thinsp;—&thinsp;I'll add the 'do not merge' label to it as well just to make sure, though.

**Comment by RandomDSdevel** [comment]
@Tristor:  Huh, looks like the ability to add labels to issues is restricted to project maintainers only.  Anyway, I submitted PR #10596 as a fix for the problems with `tesseract.rb`'s handling of `--with-opencl`.  Now we just have to wait on the resolution of tesseract-oct/tesseract#739…

**Comment by RandomDSdevel** [comment]
Hey, did you ever communicate the `autopen.sh` problems we noticed while figuring this issue out to upstream, @Tristor?

**Comment by Tristor** [comment]
Sorry, nope.  I didn't post anything in the upstream thread about this, since it looked like the core issue was addressed.  While getting a fix in the for `autopen.sh` script would be nice, it's not technically required to be fixed since it still functions okay in the current form.  You're welcome to make a PR against `tesseract` if you'd like.

**Comment by RandomDSdevel** [comment]
@Tristor:  Ah, fair enough.  I just didn't know whether or not you wanted to take care of the build script issues instead of me since you said you had run into some additional problems concerning bad build flag caching.  I'll submit the `autogen.sh` changes you proposed here in this thread upstream for review myself, then, since they look simple enough to make, and then I suppose we can leave fixing `api/Makefile.am` to someone more capable if neither of us can figure out what's wrong either between the two of us or individually.

**Comment by RandomDSdevel** [comment]
@Tristor:  OK, I submitted tesseract-ocr/tesseract#750 upstream for the `autogen.sh` issue with respect to `libtoolize`.  I mentioned your earlier [comment](https://github.com/Homebrew/homebrew-core/issues/10380#issuecomment-283506405) regarding your troubles with `$LIBS` and `$CPPFLAGS` in it, too, but won't end up surprised if the upstream devs ask us to provide more information on this latter front, either in that same pull request or in a new issue.  As you're the one who was tinkering with the build scripts more, I suspect _you're_ the one best suited to provide said information if and/or when it eventually gets requested.  In any case, I saw rumblings upstream in its issue corresponding to the initial main one here that the changes involved in fixing all these bugs might herald a new minor release, so I might have to add a release bump to #10596 before this is all said and done…

**Comment by RandomDSdevel** [comment]
@JCount and/or @Tristor:  Well, it turns out that tesseract-oct/tesseract#750 may _actually_ have been an end-user system configuration problem with respect to Tesseract's dependency, Leptonica; but I haven't quite finished confirming that just yet, as I'm waiting on a reply to [this comment](https://github.com/tesseract-ocr/tesseract/issues/739#issuecomment-287156918) on the upstream issue.

**Comment by RandomDSdevel** [comment]
Wait, no…; my local of Leptonica is linked with libtiff, so _that_ can't be the problem.  Hmmm…

**Comment by RandomDSdevel** [comment]
Folks upstream say my builds shouldn't be failing, so I'm going to try another one to see if failures might be intermittent.  If they aren't, then this _might_ actually be a _Homebrew_ bug…hmm, _grrr_, _sigh…_

**Comment by RandomDSdevel** [comment]
Well, _that_ was a waste of my time…:  my new [logs](https://github.com/RandomDSdevel/Homebrew-Installation-Logs/tree/master/Tesseract/v3.05.00/Third%20Attempt) say _exactly_ the same thing that my old ones did&thinsp;—&thinsp;module `brew update` runs and timestamps, of course.  In any case, I've now refreshed my memory as to how reproducible this bug is (100% so, and why I ever forgot _that_, I _don't_ rightly know…:man_facepalming:.)  Anyway, tesseract-ocr/tesseract#739 has been resolved to the satisfaction of its original submitter, so I'm going to have to create a new issue for the upstream equivalent of this one, which, per some comments on the now resolved issue, may well be distinct from it despite how similar it looks.

**Comment by RandomDSdevel** [comment]
OK, the new upstream issue is tesseract-ocr/tesseract#786 (though you probably already knew that due to the issue mention notification GitHub posted to this issue discussion thread above this comment…)

**Comment by RandomDSdevel** [comment]
I'll try another build once I've gotten Homebrew/formula-patches#124 and whatever ends up solving this mess out, though I currently think my near-future self _might_ balk at the thought of trying to manage multiple interdependent PRs.  Oh, well; guess I'll just have to grit my teeth and bear it…

**Comment by RandomDSdevel** [comment]
Upstream [has requested I build Tesseract outside of Homebrew](https://github.com/tesseract-ocr/tesseract/issues/786#issuecomment-289289568), so I'll attempt that soon and see what happens.

**Comment by RandomDSdevel** [comment]
Now that _think_ about it, the fact that `--with-opencl` fails for me may have something to do with the contents of the entry for my type of machine (a mid-2007 aluminum iMac) on [Apple's list of devices with OpenCL support](https://support.apple.com/en-us/HT202823).  If this is the case, maybe we should find some way to make the option error out when passed to Homebrew for use with `tesseract.rb` on machines that don't support the library?

**Comment by Tristor** [comment]
Sorry, I've been traveling so have not been able to give this any further attention.  I did want to throw out there though that I tested on a newer rMBP with dedicated graphics, so it certainly does support OpenCL and is on the list of supported devices from your link above.  Since I'm experiencing the same issue I don't know if that's the cause.

I'm on hiatus for awhile while I'm traveling but will take a look into this and any other issues I've been active on starting in June.  Thanks.

**Comment by RandomDSdevel** [comment]
@Tristor:  Yeah, you're not the only one who's reproduced this issue on a device that _does_ support OpenCL, as several folks upstream have as well.  I've almost got this issue solved now, though, as I just have to pull some patches that have appeared upstream (thanks in part to my prodding, but other contributors did most of the work, so that's partly irrelevant to an extent) back down here into Homebrew Core.  I'll do that in a new PR to unblock #10596 by the end of the day.
